### PR TITLE
Refactor instructions

### PR DESCRIPTION
**Pull request by gabi-250**
This branch contains a slightly improved version of the interpreter:
- each type of instruction corresponds to a subclass of `W_BaseInstruction` (instead of checking the instruction opcode in the main loop of the interpreter and performing the appropriate operations, the result of an instruction is obtained by calling its `execute` method)
- there are a couple of new tests (the old tests are too long and will need to be refactored as well)

### CODE REVIEW

**Comment by vext01** [comment]
The re-factoring makes sense to me and gives the interpreter a more OO design.

I notice is that we have some XXX cases which need to be addressing. I'm happy for those to be separate PRs.

Asides from that (and what Laurie has already pointed out), this looks good.

Tests pass, translation works?

**Comment by ltratt** [comment]
> Tests pass, translation works?

I wonder if there's a practical way of running the tests via Travis? Maybe we could have it download a static version of RPython and unpack it into the directory? That might be fast enough not to get chopped off at the knees by Travis.

**Comment by vext01** [comment]
That bit is easy enough. The annoying part is getting a new LLVM built on travis.

**Comment by ltratt** [comment]
Ah :/ OK, let's worry about it later.

**Comment by vext01** [comment]
Indeed ;)

File: state.py [hunk 1]
```diff
@@ -1,5 +1,62 @@
...
+        if isinstance(var, Constant):
+            return var.content
+        elif self.current_frame.has_key(var.addr):
+            return self.current_frame.get_variable(var.addr)
+        elif self.global_state and self.global_state.has_key(var.addr):
+            return self.global_state.get_variable(var.addr)
+        elif not ignore_err:
+            print "[ERROR]: Unknown variable. Exiting."
+            error_str = rffi.charp2str(llwrap.LLVMPrintValueToString(var.l_value))
+            raise NoSuchVariableException(error_str)
```
Comment (ltratt) [inline state.py:43]: Should `lookup_var` define in its description that it returns `None` if `ignore_err=False`?

**Comment by gabi-250** [comment]
Yep, it runs translated and all tests pass. I'll get to the XXXs a bit later, I still need to figure out how to fix some of them.

File: state.py (hunk 1 above)
Comment (gabi-250) [reply state.py:43]: It should return `None` if the variable is not found and `ignore_err=True`. If `ignore_err=False` and the variable is not found, it raises an exception. I'll update the description.

**Comment by vext01** [comment]
Is this ready for re-review?

**Comment by gabi-250** [comment]
Yep.

**Comment by gabi-250** [comment]
Ready.

**Comment by vext01** [comment]
@ltratt If your comments have been addressed, I am happy for this to be merged.

**Comment by ltratt** [comment]
OK, I suspect some squashing needs to happen first though?

**Comment by gabi-250** [comment]
Squashed.

**Comment by vext01** [comment]
OK, so I guess next, take a look at those XXX.

**Comment by gabi-250** [comment]
Yep.

**Pull request merged**